contracts/pallet-ibc/src/ics20/mod.rs

```diff
@@ -561,11 +563,370 @@ pub fn full_ibc_denom(packet: &Packet, mut token: PrefixedCoin) -> String {
 use ibc::applications::transfer::error::Error as Ics20Error;
 
 pub trait HandleMemo<T: Config> {
-	fn execute_memo(packet_data: &PacketData, receiver: T::AccountId) -> Result<(), Ics20Error>;
+	fn execute_memo(packet: &Packet) -> Result<(), Ics20Error>;
 }
 
 impl<T: Config> HandleMemo<T> for () {
-	fn execute_memo(_packet_data: &PacketData, _receiver: T::AccountId) -> Result<(), Ics20Error> {
+	fn execute_memo(_packet: &Packet) -> Result<(), Ics20Error> {
+		Ok(())
+	}
+}
+
+pub trait SubstrateMultihopXcmHandler {
+	type AccountId;
+
+	fn transfer_xcm(
+		from: Self::AccountId,
+		to: Self::AccountId,
+		para_id: Option<u32>,
+		amount: u128,
+		currency: u128,
+	) -> Option<()>;
+}
+
+pub struct SubstrateMultihopXcmHandlerNone<T>(PhantomData<T>);
+
+impl<T: Config> SubstrateMultihopXcmHandler for SubstrateMultihopXcmHandlerNone<T> {
+	type AccountId = T::AccountId;
+	fn transfer_xcm(
+		_: Self::AccountId,
+		_: Self::AccountId,
+		_: Option<u32>,
+		_: u128,
+		_: u128,
+	) -> Option<()> {
+		None
+	}
+}
+
+use frame_system::RawOrigin;
+use scale_info::prelude::boxed::Box;
+#[derive(serde::Serialize, serde::Deserialize, Debug)]
+pub struct Forward {
+	pub receiver: String,
+	#[serde(skip_serializing_if = "Option::is_none")]
+	pub port: Option<String>,
+	#[serde(skip_serializing_if = "Option::is_none")]
+	pub channel: Option<String>,
```

Comment:
is it possible to use `ChannelId` and `PortId` from ibc-rs?

Reply:
Channel Id and PortId in memo is a simple String and not a struct.
I am not sure how that struct will be decode, encode to be compatible with memo format

Reply:
Ok. Then you should know that the channel is not always in the format `channel-xx`, it can even be called `my-channel` so parsing [here](https://github.com/ComposableFi/centauri/pull/359/files#diff-9c98c458f1edb9cac496e225691914b2ca600002781db177e670d624083bb592R827) may fail

Reply:
But if you need this specific format, you should be able to use ChannelId and PortId, because they de/serialize to correct strings